File: Task4/src/Function.scala
import scala.collection.mutable.ArrayBuffer

/**
 * Created by Шолохов on 25.03.2015.
 */
case class Function(name: String, terms: ArrayBuffer[Expression]) extends Expression {
  override def evaluate(values: Map[String, Boolean]): Boolean = true

  override def toString(): String = {
    var s : String = name
    if (terms != null && terms.size != 0) {
      s += "("
    }
    if (terms != null && terms.size != 0) {
      terms.foreach(t => s += t + ",")
    }
    if (terms != null && terms.size != 0) {
      s += ")"
    }
    s
  }

  override def equals(o : Any): Boolean = {
    o match {
      case t : Function =>
        var res = name.equals(t.name) && terms.size == t.terms.size
        if (res) {
          terms.zip(t.terms).foreach(x => res = res && x._1.equals(x._2))
        }
        res
      case _ => false
    }
  }
}

File: Task4/src/Function.java
import java.util.ArrayList;

public class Function extends Expression {
    public final String name;
    public final ArrayList<Expression> terms;

    public Function(String name, ArrayList<Expression> terms) {
        this.name = name;
        this.terms = terms;
    }

    public boolean equals(Object other) {
        if (other instanceof Function) {
            Function fOther = (Function) other;
            if (name.equals(fOther.name) && terms.size() == fOther.terms.size()) {
                for (int i = 0; i < terms.size(); i++) {
                    if (!terms.get(i).equals(fOther.terms.get(i))) {
                        return false;
                    }
                }
                return true;
            } else {
                return false;
            }
        } else {
            return false;
        }
    }
    
    protected Expression substituteExpression(Expression where, Expression what) {
        ArrayList<Expression> newTerms = new ArrayList<>();

        for (Expression term : terms) {
            newTerms.add(term.substituteTerm(where, what));
        }

        return new Function(name, newTerms);
    }

    public String toString() {
        String str = name;
        if (terms.size() != 0) {
            str += "(";
            for (int i = 0; i < terms.size(); i++) {
                str += terms.get(i);
                if (i != terms.size() - 1) {
                    str += ", ";
                }
            }
            str += ")";
        }
        return str;
    }

    public int hashCode() {
        return toString().hashCode();
    }
}
